core/include/detray/coordinates/cylindrical3.hpp

```diff
@@ -48,6 +48,14 @@ struct cylindrical3 final : public coordinate_base<cylindrical3, transform3_t> {
         return {getter::perp(local3), getter::phi(local3), local3[2]};
     }
 
+    /** This method transforms a point from a global cartesian 3D frame to a
```

Comment:
Isn't bound only for 2D surface? Why do we need this function for 3D surfaces

Reply:
I thought bound simply meant "with respect to a reference surface'', so that the point is only valid when we have access to the surface, too, while the local one is for any point in space just in a different coordinate basis? I am using this for the grids right now, since they look up as many coordinates on a point as they have axes. So, for a 3D cylindrical system, I need three coordinates, for a 2D cylindrical system only two

Reply:
I believe we need to use `global_to_local` for 3D always.

For 2d surface, bound means the position defined on surface, and local means the position "with respect to a reference surface" (This is not the case for perigee frame though). That's why bound position is described with 2 dimension element.

For 3D coordinates, there is no surface which the position can get bound to, so we can not use the term of "bound".

Reply:
I am not completely following. The 3D shapes are used for bounding volumes and grids, not as surfaces in the propagation or detector geometry, so I am not sure why it must be a 2D point. I would argue that a bound position in this case simply means a point that is contained somewhere inside?

Anyway, I am going to rename it, that way I can also fix the cylinder2D rphi problem for the grids

Reply:
I am saying that we should not define `bound` for 3D shapes

Reply:
`global_to_local` and `global_to_bound` look to be identical functions for 3D shapes so why don't you just use `global_to_local` functions without defining new functions?

Reply:
Because I would like a single call interface somewhere, where I can say ```global_to_something``` or ```project_to_something``` for all shapes and I am sure to get back a point where all of the coordinates actually correspond to basis elements in an underlying vector space or at least to the axes of the shape. XD for an XD subspace, the way I thought coordinates are usually defined. This way, I could cleanly match the local point to the grid axes, allowing only those point types into the the search method that actually match the grid dimension.
I don't know if line surfaces have a coordinate system that can be defined this way, but there are are also no line-shaped grids needed that I know of? 

And the cylinder2D needs to be handled a bit differently in the grids apparently, so I need to make an exception for it somewhere.

Reply:
What happens if you use `global_to_local` as it used to be?

Reply:
I don't know how you can be sure that output of `global_to_bound` will be inside the 3D volume shape considering that it is same with the `global_to_local`.  I suppose, based on what you said, the point of `global_to_bound` is getting the 2D output from 2D shape and the 3D output from 3D shape?

Reply:
And where do you use 2D surface in the grid? I thought grid would be only for 3D

Reply:
Apology if I am jumping into the conclusion too quickly but I think all that confusion is coming from that we are trying to unify the interface of surface (2D) and volume (3D). and I don't see a clear reason for this unification.

Reply:
I am using 3D grids for the volume grid and probably volume material maps and 2D grids in the local navigation and surface based material maps (although a detector is in principle free to use a 3D grid for local navigation, too). So, yes, I very much would like to have a unified interface for 3D and 2D shapes, that will return 3D and 2D outputs respectively. But this will not be touching the propagation or anything in traccc, since the detector geometry is purely 2D.

A 3D grid type uses 3D local points:
```
using volume_grid = detray::grid<axes<cylinder3D>, bins::single<dindex>>;
volume_grid::point_type => 3D
```
A 2D grid, as material map on e.g. an annulus surface uses 2D local points:
```
using annulus_mat_map = detray::grid<axes<annulus2D<>>, bins::single<material_slab<>>>;
annulus_mat_map::point_type => 2D
```
Internally, this is resolved by the coordinate frame that is spanned by the grid axes, which I need in order to project a global search point onto those axes. So, for instance, for all of these different grid types, I am doing smth like this:
```
        ...
        // Track position in grid coordinates
        const auto &trf = det.transform_store()[volume.transform()];
        const auto loc_pos = local_frame().global_to_something(trf, track.pos(), track.dir());

        // Grid lookup
        return search(loc_pos, cfg.search_window);
        ...
    DETRAY_HOST_DEVICE decltype(auto) search(const point_type &p) const {
        return bin(m_axes.bins(p));
    }
```
and in a few other places, too.

Something similar happens in the bounding volume, too, which also takes on the dimension of its shape (but I did not implement it fully, yet, because we only use it in the telescope geometry building and not e.g in any search data structures, yet)

Reply:
Would it not open the possibility for some very hard to find bugs, if one could put a local point of a 2D shape in the search method of a 3D grid and then have the third element of that point be interpreted as an actual coordinate for an axis that it is not related to? I think it's better if such code does not compile any longer

Reply:
I don't know why the surface material map needs to be 2D grid... are we assuming that there won't be any material pattern along the local z-axis of surface?

Reply:
> What happens if you use `global_to_local` as it used to be?

I guess, with this PR, and 2D shapes having 2D local points, it would become impossible to put a local point of a 2D shape into a 3D grid. Right now, I expect that's possible, even though the third element in that point is not an actual coordinate

Reply:
> I don't know why the surface material map needs to be 2D grid... are we assuming that there won't be any material pattern along the local z-axis of surface?

I think that is determined by the thickness of the material slabs?

Reply:
Ah, no, I think I misunderstood the last comment. I don't think that the material will be modeled to that detail. As far as I understand the material that would get mapped onto a surface could even belong to support structures that are close to it. But I am not an expert on material mapping

Reply:
I don't think a material slab can have different materials over the local z-axis (thickness)?

Reply:
No, exactly, I misunderstood the question. As far as I know, the material in a slab is averaged

Reply:
I think you can make m_axes.bin(p) to take the first n elements of p where n is equal to the number of axes

Reply:
My main point is that I want to keep `bound` as an unique attribute of 2D surfaces.

Reply:
That is what has been happening so far. But I don't like it because a 3D grid can take a 2D local point and interpret it as a 3D local point, or am I wrong? A 3D cylinder grid could take a 2D cylinder local point and treat the radial position in the third component as a z-coordinate or not?

Reply:
> My main point is that I want to keep `bound` as an unique attribute of 2D surfaces.

OK, why can I not call it e.g. ```project_to_axes``` then?

Reply:
> a 3D grid can take a 2D local point and interpret it as a 3D local point, or am I wrong?

That can happen but is it a really problem? cartesian3 and cylindrical3 look identical to cartesian2 and cylindrical2 which is going to give you the same result in any case.

> A 3D cylinder grid could take a 2D cylinder local point and treat the radial position in the third component as a z-coordinate or not?

There is no problem with that. Radial position of cylinder is like a z position of cartesian. The only exception happens with the perigee frame

Reply:
Now I think it is possible to actually unify the 2D and 3D coordinates.

cartesian2 and cartesian3 -> cartesian
cylindrical2 and cylinderical3 -> cylindrical
polar2 -> polar
annulus2 -> annulus
line2 -> perigee

Reply:
> OK, why can I not call it e.g. project_to_axes then?

Go ahead with that If you think it is better. This will finally differentiate 2D and 3D.

Reply:
I don't know of a way to prevent that tbh and I am not saying it would become totally safe.

Anyway, what I am saying is that I would like to keep the grid design as it initially was. Nothing changes anywhere else if I add a ```project_to_axes``` to the coordinates? 

>  Radial position of cylinder is like a z position of cartesian.

I am confused: Is ```getter::perp(local3)``` a valid z-coordinate for a 3D cylinder grid?

Reply:
> > OK, why can I not call it e.g. project_to_axes then?
> 
> Go ahead with that If you think it is better. This will finally differentiate 2D and 3D.

I needed them to be different types at least to untangle the grid types

Reply:
> I am confused: Is getter::perp(local3) a valid z-coordinate for a 3D cylinder grid?

Yes that's how it works. Because when `getter::perp(local3) == radius of cylinder`, the position is bound on the cylinder surface.
Please take a look at how `reference_frame` works in cylindrical2 coordinate

Reply:
I am sorry, I don't understand that. All I am saying is, that the 3D cylinder grid needs (r, phi, z) of a given track position (which is alas not always bound to the cylinder [boundary] surface of the grid) and not (r*phi, z, r). But maybe let's discuss this at the next meeting?

Reply:
NVM, I didn't know that cylindrical3 has a different output.

Reply:
It is still true that r of cylindrical2 is used to determine whether the position is bound on surface or not.
So r of cylindrical2 and z of cartesian2 are symmetric to each other